File: js/problem24.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    if ((head === null) || (head.next === null)) {
        return head;
    }
    
    let current = head, prev = null;
    let res = head.next;
    
    while ((current !== null) && (current.next !== null)) {
        let nextPair = current.next.next;
        let temp = current.next;
        temp.next = current;
        current.next = nextPair;
        if (prev !== null) {
            prev.next = temp;
        }
        prev = current;
        current = nextPair;
    }
    
    return res;
    
};

File: leetcodeJava/src/main/java/Problem24.java
// 24. Swap Nodes in Pairs

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if ((head == null) || (head.next == null)) {
            return head;
        }
        
        ListNode current = head;
        ListNode previous = null;
        
        while (current != null) {
            ListNode temp = current.next;
            if (temp == null) {
                break;
            }
            ListNode nextPair = temp.next;
            current.next = nextPair;
            temp.next = current;
            if (previous != null) {
                previous.next = temp;
            } else {
                head = temp;
            }
            
            previous = current;
            current = nextPair;
        }
        
        return head;
    }
}